napari/utils/colormaps/standardize_color.py

```diff
@@ -0,0 +1,430 @@
+
+
+def _convert_array_to_correct_format(colors: np.ndarray) -> np.ndarray:
+    """Asserts shape, dtype and normalization of given color array.
+
+    This function deals with arrays which are already 'well-behaved',
+    i.e. have (almost) the correct number of columns and are able to represent
+    colors correctly. It then it makes sure that the array indeed has exactly
+    four columns and that its values are normalized between 0 and 1, with a
+    data type of float32.
+
+    Parameters
+    --------
+    colors : np.ndarray
+        Input color array, perhaps un-normalized and without the alpha channel.
+
+    Returns
+    ------
+    colors : np.ndarray
+        Nx4, float32 color array with values in the range [0, 1]
+    """
+    if colors.shape[1] == 3:
+        colors = np.column_stack(
+            [colors, np.ones(len(colors), dtype=np.float32)]
+        )
+
+    if colors.min() < 0:
+        raise ValueError("Colors input had negative values.")
+
+    if colors.max() > 1:
+        warnings.warn(
+            "Colors with values larger than one detected. napari"
+            " will normalize these colors for you. If you'd like to convert these"
+            " yourself, please use the proper method from skimage.color."
+        )
+        colors = _normalize_color_array(colors)
+    return np.atleast_2d(np.asarray(colors, dtype=np.float32))
+
+
+def _handle_str_list_like(colors: Sequence) -> np.ndarray:
+    """Converts lists or arrays filled with strings to the proper color array
+    format.
+
+    Parameters
+    ---------
+    colors : list-like
+        A sequence of string colors
+
+    Returns
+    ------
+    color_array : np.ndarray
+        Nx4, float32 color array
+    """
+    color_array = np.empty((len(colors), 4), dtype=np.float32)
+    for idx, c in enumerate(colors):
+        try:
+            color_array[idx, :] = _color_switch[type(c)](c)
+        except (ValueError, TypeError, KeyError):
+            raise ValueError(f"Invalid color found: {c} at index {idx}.")
+    return color_array
+
+
+def _handle_none(color) -> np.ndarray:
+    """Converts color given as None to black.
+
+    Parameters
+    --------
+    color : NoneType
+        None value given as a color
+
+    Returns
+    ------
+    arr : np.ndarray
+        1x4 numpy array of float32 zeros
+
+    """
+    return np.zeros((1, 4), dtype=np.float32)
+
+
+def _normalize_color_array(colors: np.ndarray) -> np.ndarray:
+    """Normalizes all array values in the range [0, 1].
+
+    The added complexity here stems from the fact that if a row in the given
+    array contains four identical value a simple normalization might raise a
+    division by zero exception.
+
+    Parameters
+    --------
+    colors : np.ndarray
+        A numpy array with values possibly outside the range of [0, 1]
+
+    Returns
+    ------
+    colors : np.ndarray
+        Same input array but with normalized values
```

Comment:
This function should either always copy, or be renamed `_in_place_normalize_color_array`

Reply:
Not sure what you mean, since the first line of the function uses `astype` which implicitly copies the array.

Reply:
Whoooooooaaaaaa! I've been living a lie, I thought `.astype` only copied if the types were different! I would suggest adding `.astype(np.float32, copy=True)` to prevent other clueless people from jumping to the same conclusions, but I would consider that totally optional. =)